### ISSUE TITLE
CRC32 removed?

### ISSUE THREAD

**Issue by skoshy** [sunjw/fhash#15]
Why was CRC32 removed from the latest version?

**Comment by sunjw** [comment]
Why not?

**Comment by skoshy** [comment]
I used to use CRC32 since it was very, very fast for large files, and easy to store since it's short. Was there a technical reason for removing it?

**Comment by sunjw** [comment]
For fHash, it calculates hash sums all together, CRC32 doesn't make whole process faster.
I cannot find any download provider (especially like linux distro, who is using SHA256/512 to replace SHA1) use CRC32. And myself as a very frequent user, I find no use case of CRC32.

**Comment by K0-RR** [comment]
So you removed it because you didn't use it? I also use crc because it's faster than sha.
Did removing it make the app faster, smaller or easier to develop? If not then it was a stupid decision but well, it's your app and you have any rights to do so.

**Comment by smaragdus** [comment]
For those who speak Chinese there is similar but far better tool- [MyHash](https://github.com/drag0n-app/MyHash). It is most likely abandonware but it works well. Unfortunately it is not translated into English.

**Edit - screen of MyHash:**

![MyHash 1 4 7 - 2022-05-08 - 002](https://user-images.githubusercontent.com/9025757/167273345-6a164f96-a490-4d59-be00-308739b900f3.png)

**Comment by sunjw** [comment]
> So you removed it because you didn't use it? I also use crc because it's faster than sha. Did removing it make the app faster, smaller or easier to develop? If not then it was a stupid decision but well, it's your app and you have any rights to do so.

I still don't get it. What scenario CRC32 you use for? fHash calculates all checksum together, so you get CRC32 result with other checksum. Is there any scenario that must use CRC32 and CRC32 only, and cannot use other checksum, like MD5 (which is also been supported widely and fast)?
And if you insist to use CRC32, you can still use [fHash 2.2.4](https://github.com/sunjw/fhash/releases/tag/2.2.4), which is not much different to the latest version except CRC32.

**Comment by sunjw** [comment]
> For those who speak Chinese there is similar but far better tool- [MyHash](https://github.com/drag0n-app/MyHash). It is most likely abandonware but it works well. Unfortunately it is not translated into English.
> 
> **Edit - screen of MyHash:**
> 
> ![MyHash 1 4 7 - 2022-05-08 - 002](https://user-images.githubusercontent.com/9025757/167273345-6a164f96-a490-4d59-be00-308739b900f3.png)

Which is based on very old fHash source code I think.

**Comment by smaragdus** [comment]
@sunjw 

It seems that **MyHash** is written in **Delphi**:

![MyHash 1 4 7 - 2022-05-09 - delphi - 001](https://user-images.githubusercontent.com/9025757/167328525-e0568a91-fa5e-45ec-9d6e-78e5df7b9cef.png)

The problem with **MyHash** is that the [repository](https://github.com/drag0n-app/MyHash) contains no code, just binaries. So I guess it cannot be translated and updated (without reverse engineering).

**Comment by sunjw** [comment]
> @sunjw
> 
> It seems that **MyHash** is written in **Delphi**:
> 
> ![MyHash 1 4 7 - 2022-05-09 - delphi - 001](https://user-images.githubusercontent.com/9025757/167328525-e0568a91-fa5e-45ec-9d6e-78e5df7b9cef.png)
> 
> The problem with **MyHash** is that the [repository](https://github.com/drag0n-app/MyHash) contains no code, just binaries. So I guess it cannot be translated and updated (without reverse engineering).

Very glad to know.
And also, you see, it is not an easy job to keep maintaining and improving an open (or not open) source software. I'm very happy about that there are some users using fHash and reporting issues/suggestions. But still, I consider myself as the most important user to my own software. And like fHash, I will change/improve it based on my daily usage, just like the macOS version was coming after I bought a MacBook.

The only thing I cannot understand in this issues is what kind of scenario fHash is used for, and in that scenario CRC32 is the only option. If you really like to use CRC32, you need tell the reason, which I don't have one.

**Comment by skoshy** [comment]
I personally use CRC32 for two reasons:

- It's great for detecting if there's errors with a file, particularly when downloading something to avoid corruption. According to this post, it has resistance to factors like leading/trailing zeros/spaces, etc. It's not necessarily secure, but it has a niche purpose.
- My favorite reason is It's short and can be stored easily in a database. I even store it in the filename itself because of how short it is.

Perhaps this could be a good opportunity to add options to fHash? If there was an option to enable/disable different hash functions, not only could CRC32 be re-included and be defaulted to "turned off" for people that don't need it, but others that don't need other hash functions to calculate can turn those off too.

**Comment by sunjw** [comment]
>     * It's great for detecting if there's errors with a file, particularly when downloading something to avoid corruption. According to this post, it has resistance to factors like leading/trailing zeros/spaces, etc. It's not necessarily secure, but it has a niche purpose.

But how to get the correct CRC32? You store it when it is downloaded first time? And now, you cannot verify CRC32 because fHash removed it?

> Perhaps this could be a good opportunity to add options to fHash? If there was an option to enable/disable different hash functions, not only could CRC32 be re-included and be defaulted to "turned off" for people that don't need it, but others that don't need other hash functions to calculate can turn those off too.

No, there will not have options to "turned off" some checksums.

**Comment by skoshy** [comment]
> But how to get the correct CRC32? You store it when it is downloaded first time? And now, you cannot verify CRC32 because fHash removed it?

Yes, for example if I created something I'd add the CRC32 in the filename (`project file c49516e6.xls`) and then store it on an FTP server. Then I could download it later and run the file through fHash and verify the CRC32 was the same as what I wrote originally in the filename to know it downloaded without error.

Admittedly, I've done this less compared to years ago, but I feel like I still use this feature every other month or so.
